Python 3.10 long cast bug in decompress_chunk

`decompress_chunk` [calls](https://github.com/python-lz4/python-lz4/blob/v3.1.3/lz4/frame/_frame.c#L1281) `PyArg_ParseTupleAndKeywords`, which in 3.10 for _some reason_ messes up the type of max_length (which is a `Py_ssize_t` aka `long`), and somewhere inside itself casts it to _something_ that makes it overflow. And then it's cast back to a long and the value becomes absurdly high, whic causes memory allocations to fail. Which makes the `test_frame_decompress_mem_usage` test fail.

If the value is cast to an int after `PyArg_ParseTupleAndKeywords` returns that it becomes the proper -1 value expected in the test.

While investigating this, i've came to conslusion that this _may_ be a python bug, but I can't know for certain. If you can't figure out the problem with it either, maybe it should be reported to python itself.

For completeness, the actual error with py310 (I can reproduce reliably with 3.10.0b1 via `tox -e py310`):

```pytb
__________________________________________________ test_frame_open_decompress_mem_usage[data0] ___________________________________________________

data = b'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'

    def test_frame_open_decompress_mem_usage(data):
        tracemalloc = pytest.importorskip('tracemalloc')
        tracemalloc.start()
    
        with lz4.frame.open('test.lz4', 'w') as f:
            f.write(data)
    
        prev_snapshot = None
    
        for i in range(1000):
            with lz4.frame.open('test.lz4', 'r') as f:
>               decompressed = f.read()  # noqa: F841

data       = b'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
f          = <lz4.frame.LZ4FrameFile object at 0x7f17edaab2b0>
i          = 0
prev_snapshot = None
tracemalloc = <module 'tracemalloc' from '/usr/lib/python3.10/tracemalloc.py'>

tests/frame/test_frame_5.py:80: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <lz4.frame.LZ4FrameFile object at 0x7f17edaab2b0>, size = -1

    def read(self, size=-1):
        """Read up to ``size`` uncompressed bytes from the file.
    
        If ``size`` is negative or omitted, read until ``EOF`` is reached.
        Returns ``b''`` if the file is already at ``EOF``.
    
        Args:
            size(int): If non-negative, specifies the maximum number of
                uncompressed bytes to return.
    
        Returns:
            bytes: uncompressed data
    
        """
        self._check_can_read()
>       return self._buffer.read(size)

self       = <lz4.frame.LZ4FrameFile object at 0x7f17edaab2b0>
size       = -1

lz4/frame/__init__.py:635: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_compression.DecompressReader object at 0x7f17edaabb80>

    def readall(self):
        chunks = []
        # sys.maxsize means the max length of output buffer is unlimited,
        # so that the whole input buffer can be decompressed within one
        # .decompress() call.
>       while data := self.read(sys.maxsize):

chunks     = []
self       = <_compression.DecompressReader object at 0x7f17edaabb80>

/usr/lib/python3.10/_compression.py:118: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_compression.DecompressReader object at 0x7f17edaabb80>, size = 9223372036854775807

    def read(self, size=-1):
        if size < 0:
            return self.readall()
    
        if not size or self._eof:
            return b""
        data = None  # Default if EOF is encountered
        # Depending on the input data, our call to the decompressor may not
        # return any data. In this case, try again after reading another block.
        while True:
            if self._decompressor.eof:
                rawblock = (self._decompressor.unused_data or
                            self._fp.read(BUFFER_SIZE))
                if not rawblock:
                    break
                # Continue to next stream.
                self._decompressor = self._decomp_factory(
                    **self._decomp_args)
                try:
                    data = self._decompressor.decompress(rawblock, size)
                except self._trailing_error:
                    # Trailing data isn't a valid compressed stream; ignore it.
                    break
            else:
                if self._decompressor.needs_input:
                    rawblock = self._fp.read(BUFFER_SIZE)
                    if not rawblock:
                        raise EOFError("Compressed file ended before the "
                                       "end-of-stream marker was reached")
                else:
                    rawblock = b""
>               data = self._decompressor.decompress(rawblock, size)

data       = None
rawblock   = b'\x04"M\x18@@\xc0\x0b\x01\x00\x00\x1fa\x01\x00\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xf...\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xe8Paaaaa\x00\x00\x00\x00'
self       = <_compression.DecompressReader object at 0x7f17edaabb80>
size       = 9223372036854775807

/usr/lib/python3.10/_compression.py:103: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <lz4.frame.LZ4FrameDecompressor object at 0x7f17edaa9030>
data = b'\x04"M\x18@@\xc0\x0b\x01\x00\x00\x1fa\x01\x00\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xf...\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xe8Paaaaa\x00\x00\x00\x00'
max_length = 9223372036854775807

    def decompress(self, data, max_length=-1):  # noqa: F811
        """Decompresses part or all of an LZ4 frame of compressed data.
    
        The returned data should be concatenated with the output of any
        previous calls to `decompress()`.
    
        If ``max_length`` is non-negative, returns at most ``max_length`` bytes
        of decompressed data. If this limit is reached and further output can
        be produced, the `needs_input` attribute will be set to ``False``. In
        this case, the next call to `decompress()` may provide data as
        ``b''`` to obtain more of the output. In all cases, any unconsumed data
        from previous calls will be prepended to the input data.
    
        If all of the input ``data`` was decompressed and returned (either
        because this was less than ``max_length`` bytes, or because
        ``max_length`` was negative), the `needs_input` attribute will be set
        to ``True``.
    
        If an end of frame marker is encountered in the data during
        decompression, decompression will stop at the end of the frame, and any
        data after the end of frame is available from the `unused_data`
        attribute. In this case, the `LZ4FrameDecompressor` instance is reset
        and can be used for further decompression.
    
        Args:
            data (str, bytes or buffer-compatible object): compressed data to
                decompress
    
        Keyword Args:
            max_length (int): If this is non-negative, this method returns at
                most ``max_length`` bytes of decompressed data.
    
        Returns:
            bytes: Uncompressed data
    
        """
    
        if self._unconsumed_data:
            data = self._unconsumed_data + data
    
>       decompressed, bytes_read, eoframe = decompress_chunk(
            self._context,
            data,
            max_length=max_length,
            return_bytearray=self._return_bytearray,
        )
E       MemoryError

data       = b'\x04"M\x18@@\xc0\x0b\x01\x00\x00\x1fa\x01\x00\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xf...\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xe8Paaaaa\x00\x00\x00\x00'
max_length = 9223372036854775807
self       = <lz4.frame.LZ4FrameDecompressor object at 0x7f17edaa9030>

lz4/frame/__init__.py:394: MemoryError
```

--- a/lz4/frame/__init__.py
+++ b/lz4/frame/__init__.py
@@ -617,6 +617,17 @@ class LZ4FrameFile(_compression.BaseStream):
         # returns at least one byte (except at EOF)
         return self._buffer.peek(size)
 
+    def readall(self):
+        chunks = bytearray()
+
+        while True:
+            data = self.read(io.DEFAULT_BUFFER_SIZE)
+            chunks += data
+            if not data:
+                break
+
+        return bytes(chunks)
+
     def read(self, size=-1):
         """Read up to ``size`` uncompressed bytes from the file.
 
@@ -632,6 +643,9 @@ class LZ4FrameFile(_compression.BaseStream):
 
         """
         self._check_can_read()
+
+        if size < 0 and sys.version_info >= (3, 10):
+            return self.readall()
         return self._buffer.read(size)
 
     def read1(self, size=-1):

